group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -0,0 +1,859 @@
+     *
+     * @param key   A ConsumerGroupTargetAssignmentMemberKey key.
+     * @param value A ConsumerGroupTargetAssignmentMemberValue record.
+     */
+    public void replay(
+        ConsumerGroupTargetAssignmentMemberKey key,
+        ConsumerGroupTargetAssignmentMemberValue value
+    ) {
+        String groupId = key.groupId();
+        String memberId = key.memberId();
+        ConsumerGroup consumerGroup = getOrMaybeCreateConsumerGroup(groupId, false);
+
+        if (value != null) {
+            consumerGroup.updateTargetAssignment(memberId, Assignment.fromRecord(value));
+        } else {
+            consumerGroup.removeTargetAssignment(memberId);
+        }
+    }
+
+    /**
+     * Replays ConsumerGroupTargetAssignmentMetadataKey/Value to update the hard state of
+     * the consumer group. It updates the target assignment epoch or set it to -1 to signal
+     * that it has been deleted.
+     *
+     * @param key   A ConsumerGroupTargetAssignmentMetadataKey key.
+     * @param value A ConsumerGroupTargetAssignmentMetadataValue record.
+     */
+    public void replay(
+        ConsumerGroupTargetAssignmentMetadataKey key,
+        ConsumerGroupTargetAssignmentMetadataValue value
+    ) {
+        String groupId = key.groupId();
+        ConsumerGroup consumerGroup = getOrMaybeCreateConsumerGroup(groupId, false);
+
+        if (value != null) {
+            consumerGroup.setAssignmentEpoch(value.assignmentEpoch());
+        } else {
+            if (!consumerGroup.targetAssignments().isEmpty()) {
+                throw new IllegalStateException("Received a tombstone record to delete target assignment of " + groupId
+                    + " but the assignment still has " + consumerGroup.targetAssignments().size() + " members.");
+            }
+            consumerGroup.setAssignmentEpoch(-1);
+        }
+    }
+
+    /**
+     * Replays ConsumerGroupCurrentMemberAssignmentKey/Value to update the hard state of
+     * the consumer group. It updates the assignment of a member or deletes it.
+     *
+     * @param key   A ConsumerGroupCurrentMemberAssignmentKey key.
+     * @param value A ConsumerGroupCurrentMemberAssignmentValue record.
+     */
+    public void replay(
+        ConsumerGroupCurrentMemberAssignmentKey key,
+        ConsumerGroupCurrentMemberAssignmentValue value
+    ) {
+        String groupId = key.groupId();
+        String memberId = key.memberId();
+        ConsumerGroup consumerGroup = getOrMaybeCreateConsumerGroup(groupId, false);
+        ConsumerGroupMember oldMember = consumerGroup.getOrMaybeCreateMember(memberId, false);
+
+        if (value != null) {
+            ConsumerGroupMember newMember = new ConsumerGroupMember.Builder(oldMember)
+                .updateWith(value)
+                .build();
+            consumerGroup.updateMember(newMember);
+        } else {
+            ConsumerGroupMember newMember = new ConsumerGroupMember.Builder(oldMember)
```

Comment:
would it be useful to have a helper method to delete a member?

Reply:
Yeah, I suppose I was meaning a method to add the sentinel values. But if it is only done here, maybe it is not useful.

Reply:
As it is used only once here, it is not worth it in my opinion.